borsh/tests/test_schema_conflict.rs

```diff
@@ -0,0 +1,177 @@
+    );
+    // NOTE: the contents of `defs` depend on the order of 2 above lines
+}
+
+#[test]
+fn test_implicit_conflict_slice() {
+    let mut defs = Default::default();
+    <[i64] as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+    <[ConflictingSchema] as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+    // NOTE: the contents of `defs` depend on the order of 2 above lines
+}
+
+#[test]
+fn test_implicit_conflict_array() {
+    let mut defs = Default::default();
+    <[i64; 10] as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+    <[ConflictingSchema; 10] as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+    // NOTE: the contents of `defs` depend on the order of 2 above lines
+}
+
+#[test]
+fn test_implicit_conflict_option() {
+    let mut defs = Default::default();
+    <Option<i64> as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+    <Option<ConflictingSchema> as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+    // NOTE: the contents of `defs` depend on the order of 2 above lines
+}
+
+#[allow(unused)]
+#[derive(borsh::BorshSchema)]
+struct GenericStruct<T> {
+    field: T,
+}
+
+#[test]
+fn test_implicit_conflict_struct() {
+    let mut defs = Default::default();
+    <GenericStruct<i64> as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+    <GenericStruct<ConflictingSchema> as borsh::BorshSchema>::add_definitions_recursively(
+        &mut defs,
+    );
+    // NOTE: the contents of `defs` depend on the order of 2 above lines
+    // this loophole is needed to enable derives for recursive structs/enums
+}
+
+#[allow(unused)]
+#[derive(borsh::BorshSchema)]
+struct SelfConflictingStruct {
+    field_1: i64,
+    field_2: ConflictingSchema,
+}
+
+#[test]
+#[should_panic(expected = "Redefining type schema for i64")]
+fn test_implicit_conflict_self_conflicting_struct() {
+    let mut defs = Default::default();
+    <SelfConflictingStruct as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+}
+
+#[allow(unused)]
+#[derive(borsh::BorshSchema)]
+enum GenericEnum<T> {
+    A { field: T },
+    B(u64),
+}
+
+#[test]
+fn test_implicit_conflict_enum() {
+    let mut defs = Default::default();
+    <GenericEnum<i64> as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+    <GenericEnum<ConflictingSchema> as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+    // NOTE: the contents of `defs` depend on the order of 2 above lines
+    // this loophole is needed to enable derives for recursive structs/enums
+}
+
+#[allow(unused)]
+#[derive(borsh::BorshSchema)]
+enum SelfConflictingEnum {
+    A { field: i64 },
+    B { field: ConflictingSchema },
+}
+
+#[test]
+#[should_panic(expected = "Redefining type schema for i64")]
+fn test_implicit_conflict_self_conflicting_enum() {
+    let mut defs = Default::default();
+    <SelfConflictingEnum as borsh::BorshSchema>::add_definitions_recursively(&mut defs);
+}
```

Comment:
this test passes, but if the enum definition changes to add some indirection
```rust
// SelfConflictingEnum change
#[allow(unused)]
#[derive(borsh::BorshSchema)]
enum SelfConflictingEnum {
    A { field: Vec<i64> },
    B { field: Vec<ConflictingSchema> },
}
```
the test stops to pass, unlike `test_schema_conflict` branch